### PR TITLE
Cache last anchor

### PR DESCRIPTION
**Pull request by jeffdelmerico**
For some use cases, it would be helpful to query the last anchor that was created, without needing to manually manage the anchor id.  In the case that the node is shut down and the `created_anchor` topic is no longer available, this also allows the user to recover the anchor id afterwards.  

The id is written as a string to `$HOME/.ros/last_anchor_id` by default, or to `/tmp/last_anchor_id` if `$HOME` is not defined.  

The anchor id is cached automatically, but querying this anchor on future runs of the node can be enabled by setting the `query_last_anchor_id_from_cache` parameter to true.

### CODE REVIEW

File: asa_ros/src/asa_ros_node.cpp [hunk 1]
```diff
@@ -27,20 +31,41 @@ AsaRosNode::~AsaRosNode() {}
...
+  else {
+    last_anchor_cache_path_default = "/tmp/last_anchor_id";
+  }
+  nh_private_.param<std::string>("last_anchor_cache_path",
+                                 last_anchor_cache_path_,
+                                 last_anchor_cache_path_default);
 
-  if(queue_size > 1 || use_approx_sync_policy) {
-    ROS_INFO_STREAM("Starting image and info subscribers with approximate time sync, where que-size is " << queue_size);
+  if(queue_size_ > 1 || use_approx_sync_policy_) {
```
Comment (EricVoll) [inline asa_ros/src/asa_ros_node.cpp:58]:
```suggestion
  if(use_approx_sync_policy_) {
```

Comment (EricVoll) [reply asa_ros/src/asa_ros_node.cpp:58]: I know that this code still came from my previous PR - so my mistake. The Log would say "Starting with approx. timesync" even though `use_approx_sync_policy_` might be `false`, if `queue_siye_` > 0

File: asa_ros/src/asa_ros_node.cpp [hunk 2]
```diff
@@ -260,6 +296,39 @@ void AsaRosNode::createAnchorTimerCallback(const ros::TimerEvent& e) {
   ROS_INFO_STREAM("Created an anchor with ID: " << id);
 }
 
+std::string AsaRosNode::readCachedAnchorId() {
+  std::string cached_anchor_id;
+
+  // Try to read cache file
+  std::ifstream cache_file(last_anchor_cache_path_.c_str());
+  if(cache_file) {
+    std::ostringstream ss;
```
Comment (EricVoll) [inline asa_ros/src/asa_ros_node.cpp:305]:
```suggestion
    std::ostringstream string_stream;
```

File: asa_ros/src/asa_ros_node.cpp [hunk 3]
```diff
@@ -260,6 +296,39 @@ void AsaRosNode::createAnchorTimerCallback(const ros::TimerEvent& e) {
...
+    ss << cache_file.rdbuf();
```
Comment (EricVoll) [inline asa_ros/src/asa_ros_node.cpp:306]:
```suggestion
    string_stream << cache_file.rdbuf();
```

File: asa_ros/src/asa_ros_node.cpp [hunk 4]
```diff
@@ -260,6 +296,39 @@ void AsaRosNode::createAnchorTimerCallback(const ros::TimerEvent& e) {
...
+    cached_anchor_id = ss.str();
```
Comment (EricVoll) [inline asa_ros/src/asa_ros_node.cpp:307]:
```suggestion
    cached_anchor_id = string_stream.str();
```

**Review by EricVoll — APPROVED** [review]
I made a few optional suggestions. Otherwise: LGTM

File: asa_ros/src/asa_ros_node.cpp (hunk 1 above)
Comment (jeffdelmerico) [reply asa_ros/src/asa_ros_node.cpp]: I don't remember the original intention with this, but I agree that it's not necessary here.

File: asa_ros/src/asa_ros_node.cpp (hunk 2 above)
Comment (jeffdelmerico) [reply asa_ros/src/asa_ros_node.cpp]: Good call, this is more clear.

**Pull request merged**